### ISSUE TITLE
Do not set values for arguments not provided by the user

### ISSUE THREAD

**Issue by carles-protopixel** [SupImDos/pydantic-argparse#18]
We need to know if a particular argument has been explicitly set by the user, or it is just the default value.
`pydantic` provides `model.__fields_set__` to check for those, however `pydantic-argparse`currently sets all values, regardless of the user input.

A way of doing this is by not setting defaults in `argparse` and leave `pydantic` processing those instead.

I've modified the code [in my fork](https://github.com/carles-protopixel/pydantic-argparse), and I would be happy to contribute a PR.

Thanks.

**Comment by SupImDos** [comment]
Hi @carles-protopixel 

Thanks very much for opening this issue, you raise a good point regarding the detection of unset values on `pydantic` models using `model.__fields_set__` and subsequently methods like `model.json(exclude_unset=True)`.

I've found another approach to fixing this issue, which would be to set [`argument_default=argparse.SUPPRESS`](https://docs.python.org/3/library/argparse.html#argument-default) on the underlying internal `argparse.ArgumentParser` and to remove the defaults on all of the optional arguments. From the `argparse` documentation:

> Providing default=argparse.SUPPRESS causes no attribute to be added if the command-line argument was not present.

This should cause optional arguments to be ignored (and not set on the internal `argparse.Namespace`) when they are not supplied, so that `pydantic` can handle the setting of the defaults itself.

I've made a new branch with this change here: [feature/exclude-unset-defaults](https://github.com/SupImDos/pydantic-argparse/tree/feature/exclude-unset-defaults). If you would like to see the code diff, the commit for this change is here: b2ef192894651e2e7266523429b6c414fda27d45.

Consider the following code-snippet:

```python
import pydantic
import pydantic_argparse


class Arguments(pydantic.BaseModel):
    string: str = pydantic.Field(description="a required string")
    integer: int = pydantic.Field(42, description="an optional integer")


def main() -> None:
    # Create Parser and Parse Args
    parser = pydantic_argparse.ArgumentParser(
        model=Arguments,
        prog="Example Program",
        description="Example Description",
        version="0.0.1",
        epilog="Example Epilog",
    )
    args = parser.parse_typed_args()

    # Print Args
    print("JSON (Exclude Unset = False):", args.json(exclude_unset=False))
    print("JSON (Exclude Unset =  True):", args.json(exclude_unset=True))


if __name__ == "__main__":
    main()
```

The output for the above on `v0.5.0`:
```shell
$ python3 example.py --string hello             
JSON (Exclude Unset = False): {"string": "hello", "integer": 42}
JSON (Exclude Unset =  True): {"string": "hello", "integer": 42}  # Wrong!!!

$ python3 example.py --string hello --integer 42
JSON (Exclude Unset = False): {"string": "hello", "integer": 42}
JSON (Exclude Unset =  True): {"string": "hello", "integer": 42}

$ python3 example.py --string hello --integer 97
JSON (Exclude Unset = False): {"string": "hello", "integer": 97}
JSON (Exclude Unset =  True): {"string": "hello", "integer": 97}
```

The output for the above on `feature/exclude-unset-defaults`:
```shell
$ python3 example.py --string hello
JSON (Exclude Unset = False): {"string": "hello", "integer": 42}
JSON (Exclude Unset =  True): {"string": "hello"}

$ python3 example.py --string hello --integer 42
JSON (Exclude Unset = False): {"string": "hello", "integer": 42}
JSON (Exclude Unset =  True): {"string": "hello", "integer": 42}

$ python3 example.py --string hello --integer 97
JSON (Exclude Unset = False): {"string": "hello", "integer": 97}
JSON (Exclude Unset =  True): {"string": "hello", "integer": 97}
```

Does this look like the behaviour you are looking for?

All of the lints, type-checks and unit tests still pass for me (on `3.10.7`) with this change. However, I did find a possible issue (https://github.com/python/cpython/pull/30111) in Python's `argparse` related to the usage of `argparse.BooleanOptionalAction` with `argparse.SUPPRESS`. This *could* possibly effect users on affected versions of Python who also use required boolean arguments (maybe?). I'm not sure how the backports of the bugfix work, but we may need to introduce some handling for this bug if you (or other users) experience an issue.

Cheers!

**Comment by SupImDos** [comment]
Quick update on my comment above regarding the possible `argparse` bug in some versions of Python:

> All of the lints, type-checks and unit tests still pass for me (on `3.10.7`) with this change. However, I did find a possible issue ([python/cpython#30111](https://github.com/python/cpython/pull/30111)) in Python's `argparse` related to the usage of `argparse.BooleanOptionalAction` with `argparse.SUPPRESS`. This _could_ possibly effect users on affected versions of Python who also use required boolean arguments (maybe?). I'm not sure how the backports of the bugfix work, but we may need to introduce some handling for this bug if you (or other users) experience an issue.

I seem to have found a simple 1-line workaround for the above bug. We should be able to just specify `default=None` on the required boolean arguments (see: d9c00bbc21f05b36136458039423ce56e0707a09). The default value is never used, but it allows us to bypass the missing check for `default is argparse.SUPPRESS` in the help string construction.

**Comment by carles-protopixel** [comment]
The approach you mention seems fine to me. Maybe it makes sense to detect the affected python versions and issue a warning?

I am currently using a Python version affected by the bug (3.9.10), but it seems reasonable to me to upgrade.
In any case, thank you very much!

**Comment by SupImDos** [comment]
Closed in #21, released in [`v0.6.0`](https://github.com/SupImDos/pydantic-argparse/releases/tag/v0.6.0) ([PyPI](https://pypi.org/project/pydantic-argparse/0.6.0/))
